src/lightbox/downloadImage.js

```diff
@@ -35,27 +34,35 @@ const androidEnsureStoragePermission = async (): Promise<void> => {
 };
 
 export default async (src: string, auth: Auth): Promise<mixed> => {
-  const absoluteUrl = getFullUrl(src, auth.realm);
+  const fileDownloadUrl: string = await getFileDownloadUrl(src, auth);
 
   if (Platform.OS === 'ios') {
-    const delimiter = absoluteUrl.includes('?') ? '&' : '?';
-    const urlWithApiKey = `${absoluteUrl}${delimiter}api_key=${auth.apiKey}`;
-    return CameraRoll.saveToCameraRoll(urlWithApiKey);
+    try {
+      return CameraRoll.saveToCameraRoll(fileDownloadUrl);
```

Comment:
In order for this to work as expected, we'll need to be sure that the `Promise` returned by `CameraRoll.saveToCameraRoll` rejects when we want it to reject, i.e., when we want to open the image in the browser.

I haven't actually found anything in the [doc](https://reactnative.dev/docs/0.59/cameraroll) that says the promise will *ever* reject. If I remember correctly, the failure over [here](https://github.com/zulip/zulip-mobile/pull/4089#discussion_r421178576) (dealing with a `.docx` file instead of an image) was in fact that the promise never resolved *or* rejected, it just sat there. Hmm, that wasn't a helpful behavior in that case, and it wouldn't be helpful here either.

Looking back at the commit message, partly to remind myself what failure we're aiming to handle (emphasis mine):

> If we fail to download the file, **which happens if we failed to**
> **generate the temporary URL because the server does not support it**,
> we open it in the browser instead,

It looks like the really crucial bit is whether we failed to generate the temporary authless URL (corresponding to whether the server supports it) or not. With that question answered, we don't need try/catch logic around the actual file download (either on iOS or Android).

Hmm, but `getFileDownloadUrl` kind of swallows the direct answer to that question. The answer is perfectly clear in the *body* of `getFileDownloadUrl`, but a string is returned in either case.

- I guess we could check `fileDownloadUrl` against `getFullUrl(src, auth)`, but this seems indirect.
- It's also a bit weird, but maybe `getFileDownloadUrl` should return an object holding the same URL it's been computing, plus a boolean value that tells us if the endpoint was used successfully to generate the URL?

I'm not sure; I think I'd lean toward the latter. @gnprice, maybe you have a preference? Whichever way we go, it'll be good to mention in a code comment why it's necessary to act differently when using the regular URL or the temporary one from the endpoint.

And I think avoiding the potentially confusing (and even buggy, if that promise never resolves or rejects!) try/catch logic around `CameraRoll.saveToCameraRoll` (and the counterpart for Android) will be really helpful.

Reply:
> * I guess we could examine that string, by checking `fileDownloadUrl` against `getFullUrl(src, auth)`, but this seems indirect and error-prone if we have to do it at many call sites.

I agree.


> * maybe `getFileDownloadUrl` should return an object holding the same URL it's been computing, plus a boolean value that tells us if the endpoint was used successfully to generate the URL?

Hmm let's see. As I'm reading this, the idea is that
* we want to use the generated download URL when we can;
* sometimes we can't and then we want to just use the plain absolute URL (from `getFullUrl`);
* and at the other call site, whichever URL we use we're going to do the same thing with it,
* but at this call site, we're going to do one thing with the generated URL (i.e. directly download it) and another thing with the fallback URL (i.e. open in a browser).

Is that right?

I think the cleanest way to handle that is probably: we want part of the control flow to be different between the call sites, so push that part out into the call sites. Like:
* have `tryGetTemporaryFileUrl: (string, Auth) => Promise<null | string>`, which just returns the API result if applicable and `null` if not;
* then at the call sites say something like
```js
const tempUrl = tryGetTemporaryFileUrl(src, auth);
if (tempUrl !== null) {
  // ... use tempUrl ...
  return;
}
const fullUrl = getFullUrl(src, auth.realm);
// ... use fullUrl ...
```
* oh and in the case where both kinds of URLs will be treated the same, that can be simplified to
```js
const url = tryGetTemporaryFileUrl(src, auth) ?? getFullUrl(src, auth.realm);
// ... use url ...
```

Reply:
Used this approach.